### PR TITLE
Add `output_file` and `reporter` option to `swiftlint` action

### PR DESCRIPTION
**Pull request by noboru-i**
I hope to receive SwiftLint result that is formatted JSON.
Sample is here https://github.com/noboru-i/kyouen-ios/blob/bcbca4f21be08d56242e23a44f04986496515bd3/fastlane/Fastfile#L80

SwiftLint can switch reporter by using `.swiftlint.yml`.
So I make backup and replace `reporter` when executing `swiftlint`.

Sorry for less test code.
I don't know how to test that ".swiftlint.yml" is correct when executing `swiftlint`.

### CODE REVIEW

File: lib/fastlane/actions/swiftlint.rb [hunk 1]
```diff
@@ -21,6 +55,16 @@ def self.details
...
         [
+          FastlaneCore::ConfigItem.new(key: :output_file,
+                                       description: 'Path to output SwiftLint result',
+                                       optional: true),
+          FastlaneCore::ConfigItem.new(key: :reporter,
+                                       description: 'Custom reporter of SwiftLint result',
+                                       optional: true,
+                                       default_value: 'xcode',
+                                       verify_block: proc do |value|
+                                         fail 'Unknown reporter' unless VALID_REPORTERS.include?(value)
```
Comment (KrauseFx) [inline lib/fastlane/actions/swiftlint.rb]: Can you use `raise` instead of `fail`

File: lib/fastlane/actions/swiftlint.rb [hunk 2]
```diff
@@ -1,11 +1,45 @@
...
+        command = 'swiftlint'
+        command << " > #{params[:output_file]}" if params[:output_file]
+        Actions.sh(command)
+
+      ensure
+        restore(config_file)
+      end
+
+      def self.backup(config_file)
+        FileUtils.cp(config_file, "#{config_file}.back", preserve: true) if File.exist? config_file
```
Comment (KrauseFx) [inline lib/fastlane/actions/swiftlint.rb]: Why do we do a backup and restore? Could you add some comments describing this

Comment (noboru-i) [inline lib/fastlane/actions/swiftlint.rb]: First reason is below,
> SwiftLint can switch reporter by using .swiftlint.yml.
> So I make backup and replace reporter when executing swiftlint.

Second reason is,
I hope to no diff between executing `swiftlint` before / after.
If some changes occured, make some noise when hitting `git status`.
I got better idea is `backup and restore`.

**Comment by noboru-i** [comment]
@KrauseFx Thank you for quick response.
I fix code and describe why need backup.
Please check it again.

I check TravisCI, failed reason is below,

> No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.

I think, we need to retry build. But I cannot find retry button for this repository.

**Comment by mfurtak** [comment]
Hi @noboru-i, thank you for your contribution!

Can you explain a little more about why you need to use the fastlane action to temporarily change the contents of the `.swiftlint.yml` configuration file? Is it not possible for you to keep these configuration values in your `.swiftlint.yml` in your source control instead?

**Comment by noboru-i** [comment]
@mfurtak Thank you for replying.

Sorry, I'm not good at English...
You mean is like below?

1. Don't create backup.
2. Change the contents of the `.swiftlint.yml`.
3. Execute `swiftlint`.
4. Execute `git checkout .swiftlint.yml`.

In above, I afraid of rolling back the editing changes like below,

1. Edit `.swiftlint.yml` in local. (But don't commit it.)
2. Execute fastlane include above swiftlint action.
3. Rolling back the `.swiftlint.yml` that before step 1.

I hope to keep edited the `.swiftlint.yml` after step 1.

---

I have plan B.
I remove the `reporter` option and remove backup/restore process in this pull request.
I can switch reporter by editing `Fastfile` like below,

```
  lane :lint do
    backup_file(path: '.swiftlint.yml')

    require 'yaml'
    lint_config = YAML.load_file('../.swiftlint.yml')
    puts lint_config
    lint_config['reporter'] = 'json'
    open '../.swiftlint.yml', 'w' do |f|
      YAML.dump(lint_config, f)
    end

    swiftlint(
      output_file: 'swiftlint.result.json'
    )

    restore_file(path: '.swiftlint.yml')
  end
```

If you hope the plain B, I will edit code for it.

**Comment by mfurtak** [comment]
No problem, your English is quite good! :+1: I am sorry for not making my question more clear.

If you want to have `reporter: json` in your `.swiftlint.yml` configuration when you run the `swiftlint` command, why not have that configuration file contain this option all the time? Is there a reason you need to put this value in, and then take it back out by using backup and restore?

Basically I'm trying to understand if this is a common problem for using the `swiftlint` tool. Thanks! :smile:

**Comment by noboru-i** [comment]
@mfurtak Thank you for describing more.
I may understand you wanted answer.

I want to have `reporter: xcode` in the `.swiftlint.yml` when developing code.
When setting https://github.com/realm/SwiftLint#xcode , Xcode want that reporter formatted result.

But I want to have `reporter: json` when executed in CI.
I created https://github.com/noboru-i/swiftlint_translate_checkstyle_format that read json formatted result.

So I want to have `reporter: json` when executing `swiftlint` by only fastlane.

**Comment by mfurtak** [comment]
OK, I see! Thank you. That explanation is very helpful.

From what I can tell, SwiftLint supports choosing a different configuration file using the `--config` option. What if we add an option to our fastlane command to allow users to pass the name of the desired config file?

Then, you can have two configuration files in your project: `.swiftlint.yml` for normal development, and `.swiftlint-ci.yml` with json reporting settings for running on CI. This way both files can be exactly as desired all the time, and there is no need to worry about backup/restore or git diff.

Would that be an acceptable solution?

**Comment by noboru-i** [comment]
Thank you for suggestion!
That is great idea. I didn't think it.

I try to use the `--config` option.
And delete backup/restore process.

**Comment by noboru-i** [comment]
@mfurtak 
I fix this pull request for using `--config` instead of backup/restore action.
That is nice solution for me and makes the code smart.

Please review this. Thank you!

**Comment by mfurtak** [comment]
@noboru-i This looks great! Please squash it to one commit, and we will get it merged. :+1:

**Comment by noboru-i** [comment]
@mfurtak Thank you for reviewing.
I squashed it to one commit.

**Comment by mfurtak** [comment]
Thank you for your contribution! :rocket:

**Comment by noboru-i** [comment]
Thank you! :+1:

**Pull request merged**